### ISSUE TITLE
Confusing keybinding error and behaviour

### ISSUE THREAD

**Issue by Akuukis** [awesomeWM/awesome#2147]
Output of `awesome --version`:
```
awesome v4.2 (Human after all)
 • Compiled against Lua 5.3.3 (running with Lua 5.3)
 • D-Bus support: ✔
 • execinfo support: ✔
 • xcb-randr version: 1.5
 • LGI version: 0.9.2
```

**How to reproduce the issue:**

1. Have v4.0 `rc.lua` with some keybindings to "PageUp" or "PageDown" keys, i.e. `awful.key({ MODKEY,           }, "PageUp"  , function() awful.screen.focus_relative( 1) end, {description = "focus next screen", group = "navigation"}),`.
2. It is working on laptop No.1 that has big keyboard with "PageUp" and "PageDown" buttons on it.
3. Install Awesome on laptop No. 2 that doesn't have big keyboard with "PageUp" and "PageDown" buttons on it (instead they are under CTRL+FN+UP/DOWN).
4. Copy config over from Laptop No.1 to laptop No.2

**Actual result:**

Awesome fails to load `rc.lua` and throws error:
```
error while running function!
stack traceback:
    [C]: in field 'key'
    /usr/share/awesome/lib/awful/key.lua:88: in function 'awful.key.new'
    (...tail calls...)
    /home/<$USER>/.config/awesome/rc.lua:223: in main chunk
error: invalid key to 'next'
```
where rc.lua:223 is the line in step No.1.

**Expected result:**

Either,
* error that tells something along the lines of *provided key "PageUp" does not exist but it has to, so please use `xev` or similar utility to find the correct name of it*
* warning that given keybinding will be effectively void because it is impossible to ever press a key which does not exist on current keyboard layouts. Unless keyboards and/or layouts are changed.

In the same `rc.lua` I had another keybinding to "Space" that had to be renamed to " ". I would be happy to see the same expected behavior for that case also.

**Continuing (probably should have its own issue)**

I do have external keyboard with "PageUp" (112) and "PageDown" (117) keys on it. After I add that to my laptop via usb, still neither keyname nor keycode works and rc.lua keeps failing to load. Mainly I am working from the big external keyboard, but how can I now use those keys in keybindings with the new laptop?

Also, pressing CTRL+FN+UP on my small keyboard successfully produces keypress with keycode 112 that can be registered by `xov`. Probably that's lower hardware level, I don't know. Nevertheless, technically my keyboard is able to produce keypress with keycode 112, thus should be ok, right?

**Comment by psychon** [comment]
```
error while running function!
stack traceback:
    [C]: in field 'key'
    /usr/share/awesome/lib/awful/key.lua:88: in function 'awful.key.new'
    (...tail calls...)
    /home/<$USER>/.config/awesome/rc.lua:223: in main chunk
error: invalid key to 'next'
```
I'm confused... This is clearly the call to the CAPI function to create a new key object. However, I do not see any call to `next` in there...
What is your `MODKEY` set to exactly? Do you modify `awful.key.ignore_modifiers` in your config? I don't think that this error has anything to do with "PageUp".
Does running `key({ modifiers = { MODKEY }, key = "PageUp" })` produce the same error?
If not, does one of the following calls produce the error? Which one?
```
local subsets = require("gears.math").subsets(awful.key.ignore_modifiers)
for _, set in ipairs(subsets) do
   print("Before", require("gears.debug").dump_return(set))
   key({modifiers = require("gears.table").join({ MODKEY }, set), key = "PageUp"})
end
```

**Comment by Akuukis** [comment]
> What is your MODKEY set to exactly?

default: `local MODKEY = "Mod4"`. I do have such key on my small keyboard.

> Do you modify `awful.key.ignore_modifiers` in your config?

No

> I don't think that this error has anything to do with "PageUp".

Deleting those lines with "PageUp" and "PageDown" solves the error, except that I don't have a keybinding anymore. Also changing keyname to something else, i.e. "a", solves the error.

> Does running `key({ modifiers = { MODKEY }, key = "PageUp" })` produce the same error?

I put the line in `rc.lua` after my modkey definition on line 9, reload and get the following error:
```
error while running function!
stack traceback:
    [C]: in global 'key'
    /home/<$USER>/.config/awesome/rc.lua:9: in main chunk
error: invalid key to 'next'
```

> If not, does one of the following calls produce the error? Which one?

It produces the same error, but I dont know for which one. Where can I look at stdout?

**Comment by psychon** [comment]
Weird... no idea what is going on here.
Does the following also reproduce the problem?
```
key({modifiers = { MODKEY }}).key = "PageUp"
```
The error seems to come from here: https://sources.debian.org/src/lua5.3/5.3.3-1/src/ltable.c/?hl=181#L181
Which means that something called the lua function `next` with a key that is not present on the table. I don't understand where this comes from, nor what this has to do with `PageUp`. Binding that key should work even if it is not present on the keyboard?!

**Comment by Akuukis** [comment]
Thank you for looking into it, I appreciate your help! Unfortunately I don't have any clue myself about what is going on, so don't know how to help here.

> Does the following also reproduce the problem? `key({modifiers = { MODKEY }}).key = "PageUp"`

Added to `rc.lua` & reloaded --> no errors whatsoever and `rc.lua` loaded successfully.

**Comment by hoelzro** [comment]
@psychon That error usually corresponds to a situation in which a key is added during iteration using `pairs` or `next - from https://www.lua.org/manual/5.3/manual.html#pdf-next:

> The behavior of next is undefined if, during the traversal, you assign any value to a non-existent field in the table. You may however modify existing fields. In particular, you may clear existing fields.

**Comment by psychon** [comment]
@hoelzro Yup, I know. But the only call to `lua_next` in `objects/key.c` is:
```
    lua_pushnil(L);
    while(lua_next(L, idx))
        if(luaA_toudata(L, -1, &key_class))
            key_array_append(keys, luaA_object_ref_item(L, oidx, -1));
        else
            lua_pop(L, 1);
```
However, this code is only called from `c:keys(foo)` (so not here) and it looks correct to me.

We already established that this comes from `key({ modifiers = { MODKEY }, key = "PageUp" })` and that `key({modifiers = { MODKEY }}).key = "PageUp"` does not reproduce the problem. The code that handles `key = ` is
```c
static int
luaA_key_set_key(lua_State *L, keyb_t *k)
{
    size_t klen;
    const char *key = luaL_checklstring(L, -1, &klen);
    luaA_keystore(L, -3, key, klen);
    return 0;
}
```
This might convert the value to a string if you do `key = 42`, but that's clearly not the problem here. This function is called from here:
```c
int
luaA_class_new(lua_State *L, lua_class_t *lua_class)
{
    /* Check we have a table that should contains some properties */
    luaA_checktable(L, 2);

    /* Create a new object */
    void *object = lua_class->allocator(L);

    /* Push the first key before iterating */
    lua_pushnil(L);
    /* Iterate over the property keys */
    while(lua_next(L, 2))
    {
        /* Check that the key is a string.
         * We cannot call tostring blindly or Lua will convert a key that is a
         * number TO A STRING, confusing lua_next() */
        if(lua_isstring(L, -2))
        {
            lua_class_property_t *prop = luaA_class_property_get(L, lua_class, -2);

            if(prop && prop->new)
                prop->new(L, object);
        }
        /* Remove value */
        lua_pop(L, 1);
    }

    return 1;
}
```
which also looks correct to me.

So, as far as I see, nothing modifies the table during traversal, and nothing converts e.g. an integer to a string.

**Comment by psychon** [comment]
> awesome v4.2 (Human after all)
>  • Compiled against Lua 5.3.3 (running with Lua 5.3)

Hmm... This might be #2106 which was fixed by e20068cb4a781b7bb2f384fecee9bb6aa80de7e7 (`luaA_warn` keeps a traceback on the stack; since `lua_next` takes the value ontop of the stack as key, this would mean that it tries to continue iteration with the traceback).

@Akuukis Do you have any warnings on awesome's stderr? You *might* be able to figure out where awesome's stderr goes to via `ls -l /proc/$(pidof awesome)/fd/2`. You could then do `awesome-client 'key({ modifiers = { MODKEY }, key = "PageUp" })'` to cause this problem and then check if a new warning appeared in the output.

**Comment by Akuukis** [comment]
@psychon 

Writing `awesome-client 'key({ modifiers = { MODKEY }, key = "PageUp" })'` into console produces red popup with 
> ### Oops, an error happened!
> invalid key to 'next'

Command `ls -l /proc/$(pidof awesome)/fd/2` points me to unix socket which I don't know what to do about. Tried using `socat` but without success, any suggestions?

**Comment by Akuukis** [comment]
@actionless Yes, I tried one that should be equivalent. Now I try your command but get the same result.
```
$ socat /proc/$(pidof awesome)/fd/2 -
2018/04/11 18:13:00 socat[14962] E connect(5, AF=1 "/proc/1410/fd/2", 17): Connection refused
$ socat /proc/$(pidof awesome)/fd/2 /dev/stdout
2018/04/11 18:13:00 socat[14962] E connect(5, AF=1 "/proc/1410/fd/2", 17): Connection refused
```

**Comment by Akuukis** [comment]
I am using Gnome Login Screen that was by default with Debian 9.4, then configured awesome with `NoDisplay=false` to show up in window manager list.

I read in the awesome FAQ the same you asked me, and we know the socket, but no idea how to access that.

**Comment by Akuukis** [comment]
Thanks @actionless ! I played with your script few times but didn't manage to run it. I have gdm, and when I wrote executable to this script within `/usr/share/xsessions/awesome.desktop` gdm kept crashing without starting awesome - again nowhere to find logs.

**Comment by actionless** [comment]
weird, mb you didn't set script to be executable (chmod +x) or don't have ~/.cache directory (my bad, i don't have a `mkdir -p` in the script)

**Comment by mstensho** [comment]
I stubled into the same problem. Try "Page_Up" and "Page_Down" instead. Works for me.

awesome v4.2 (Human after all)
 • Compiled against Lua 5.3.3 (running with Lua 5.3)
 • D-Bus support: ✔
 • execinfo support: ✔
 • xcb-randr version: 1.5
 • LGI version: 0.9.2
